core/modules/wcontrol/TransmitMgr.h

```diff
@@ -0,0 +1,119 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_WCONTROL_TRANSMITMGR_H
+#define LSST_QSERV_WCONTROL_TRANSMITMGR_H
+
+// System headers
+#include <assert.h>
+#include <atomic>
+#include <condition_variable>
+#include <mutex>
+
+// Qserv headers
+
+
+namespace lsst {
+namespace qserv {
+namespace wcontrol {
+
+class TransmitLock;
+/// A way to limit the number of concurrent
+/// transmits. 'interactive queries' are not blocked.
+/// New tasks cannot transmit to the czar until the number of jobs
+/// currently transmitting data drops below maxAlreadyTran
+/// Priority is given to finish tasks that have already started transmitting.
+/// This class is meant to improve transmission efficiency, and
+/// it is possible for _alreadyTransCount to go above normal
+/// limits due to a race condition. Fixing the race condition may well
+/// slow things down more than the occasional breach of the
+/// _alreadyTransCount limit.
+/// TODO:
+///    -- The czar these are being sent to should be taken into consideration
+///       as the limit should really be per czar.
+///    -- maybe change at runtime.
+class TransmitMgr {
+public:
+    using Ptr = std::shared_ptr<TransmitMgr>;
+
+    TransmitMgr(int maxTransmits, int maxAlreadyTran)
+        : _maxTransmits(maxTransmits),  _maxAlreadyTran(maxAlreadyTran) {
+        assert(_maxTransmits > 1);
+        assert(_maxAlreadyTran > 1);
+    }
+    TransmitMgr() = delete;
+    TransmitMgr(TransmitMgr const&) = delete;
+    TransmitMgr& operator=(TransmitMgr const&) = delete;
+    virtual ~TransmitMgr() = default;
+
+    int getTotalCount() { return _totalCount; }
+    int getTransmitCount() { return _transmitCount; }
+    int getAlreadyTransCount() { return _alreadyTransCount; }
+
+    virtual std::ostream& dump(std::ostream &os) const;
+    std::string dump() const;
+    friend std::ostream& operator<<(std::ostream &out, TransmitMgr const& mgr);
+
+    friend class TransmitLock;
+
+private:
+    void _take(bool interactive, bool alreadyTransmitting);
+
+    void _release(bool interactive, bool alreadyTransmitting);
+
+    std::atomic<int> _totalCount{0};
+    std::atomic<int> _transmitCount{0};
+    std::atomic<int> _alreadyTransCount{0};
+    int _maxTransmits;
+    int _maxAlreadyTran;
+    std::mutex _mtx;
+    std::condition_variable _tCv;
+};
+
+
+/// RAII class to support TransmitMgr
+class TransmitLock {
+public:
+    TransmitLock(TransmitMgr& transmitMgr, bool interactive, bool alreadyTransmitting)
```

Comment:
Efficiency, easier to read. A RAII lock should have pretty limited scope, so incrementing the values in a shared pointer will just slow things down more by locking another mutex. Holding a reference to a shared pointer (instead of incrementing the pointer) has all the same risks as the reference.